spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -58,53 +61,138 @@ case class RewriteMergeInto(conf: SQLConf) extends Rule[LogicalPlan] with Rewrit
-        // TODO : Optimize this to use LEFT ANTI or RIGHT OUTER when applicable.
+
+        val (mergeBuilder, targetTableScan) = buildDynamicFilterTargetScan(target, source, cond)
+
+        // rewrite the matched actions to ensure there is always an action to produce the output row
+        val (matchedConditions, matchedOutputs) = rewriteMatchedActions(matchedActions, target.output)
+
+        // use a full outer join because there are both matched and not matched actions
         val sourceTableProj = source.output ++ Seq(Alias(TRUE_LITERAL, ROW_FROM_SOURCE)())
-        val targetTableProj = target.output ++ Seq(Alias(TRUE_LITERAL, ROW_FROM_TARGET)())
+        val targetTableProj = targetTableScan.output ++ Seq(Alias(TRUE_LITERAL, ROW_FROM_TARGET)())
         val newTargetTableScan = Project(targetTableProj, targetTableScan)
         val newSourceTableScan = Project(sourceTableProj, source)
         val joinPlan = Join(newSourceTableScan, newTargetTableScan, FullOuter, Some(cond), JoinHint.NONE)
 
-        // Construct the plan to replace the data based on the output of `MergeInto`
         val mergeParams = MergeIntoParams(
           isSourceRowNotPresent = IsNull(findOutputAttr(joinPlan.output, ROW_FROM_SOURCE)),
           isTargetRowNotPresent = IsNull(findOutputAttr(joinPlan.output, ROW_FROM_TARGET)),
-          matchedConditions = matchedActions.map(getClauseCondition),
-          matchedOutputs = matchedActions.map(actionOutput(_, target.output)),
+          matchedConditions = matchedConditions,
+          matchedOutputs = matchedOutputs,
           notMatchedConditions = notMatchedActions.map(getClauseCondition),
-          notMatchedOutputs = notMatchedActions.map(actionOutput(_, target.output)),
-          targetOutput = target.output :+ FALSE_LITERAL,
-          deleteOutput = target.output :+ TRUE_LITERAL,
+          notMatchedOutputs = notMatchedActions.map(actionOutput),
+          targetOutput = target.output,
           joinedAttributes = joinPlan.output
         )
         val mergePlan = MergeInto(mergeParams, target, joinPlan)
         val batchWrite = mergeBuilder.asWriteBuilder.buildForBatch()
+
         ReplaceData(target, batchWrite, mergePlan)
     }
   }
 
-  private def actionOutput(clause: MergeAction, targetOutputCols: Seq[Expression]): Seq[Expression] = {
+  private def actionOutput(clause: MergeAction): Option[Seq[Expression]] = {
     clause match {
       case u: UpdateAction =>
-        u.assignments.map(_.value) :+ FALSE_LITERAL
+        Some(u.assignments.map(_.value))
       case _: DeleteAction =>
-        targetOutputCols :+ TRUE_LITERAL
+        None
       case i: InsertAction =>
-        i.assignments.map(_.value) :+ FALSE_LITERAL
+        Some(i.assignments.map(_.value))
     }
   }
 
   private def getClauseCondition(clause: MergeAction): Expression = {
     clause.condition.getOrElse(TRUE_LITERAL)
   }
+
+  private def buildDynamicFilterTargetScan(
```

Comment:
It feels like it would be more readable not to bundle the creation of `MergeBuilder` in this method.

Reply:
This is just to avoid duplication above between the full outer and right outer join cases. I can remove it, but it looked like it needed to be refactored before.